File: 2023/day19/perl/solution.pl
#!/usr/bin/env perl
# Day 19: Aplenty - Workflow processing and range analysis
use strict;
use warnings;
use File::Basename;

my $dir = dirname(__FILE__);
my $input_file = "$dir/../input.txt";

# Parse input
sub parse_input {
    my ($filename) = @_;
    open my $fh, '<', $filename or die "Cannot open $filename: $!";
    local $/;
    my $text = <$fh>;
    close $fh;

    $text =~ s/\s+$//;
    my ($workflow_section, $parts_section) = split /\n\n/, $text;

    # Parse workflows
    my %workflows;
    for my $line (split /\n/, $workflow_section) {
        my ($name, $rules_str) = $line =~ /^(\w+)\{(.+)\}$/;
        my @rules;
        for my $rule (split /,/, $rules_str) {
            if ($rule =~ /:/) {
                my ($condition, $destination) = split /:/, $rule;
                my ($attr, $op, $value) = $condition =~ /^([xmas])([<>])(\d+)$/;
                push @rules, [$attr, $op, int($value), $destination];
            } else {
                # Default rule
                push @rules, [undef, undef, undef, $rule];
            }
        }
        $workflows{$name} = \@rules;
    }

    # Parse parts
    my @parts;
    for my $line (split /\n/, $parts_section) {
        my %part;
        while ($line =~ /([xmas])=(\d+)/g) {
            $part{$1} = int($2);
        }
        push @parts, \%part;
    }

    return (\%workflows, \@parts);
}

# Process a single part through workflows
sub process_part {
    my ($workflows, $part) = @_;
    my $current = 'in';

    while ($current ne 'A' && $current ne 'R') {
        for my $rule (@{$workflows->{$current}}) {
            my ($attr, $op, $value, $destination) = @$rule;
            if (!defined $attr) {
                # Default rule
                $current = $destination;
                last;
            } elsif ($op eq '<' && $part->{$attr} < $value) {
                $current = $destination;
                last;
            } elsif ($op eq '>' && $part->{$attr} > $value) {
                $current = $destination;
                last;
            }
        }
    }

    return $current eq 'A';
}

# Part 1: Sum ratings of accepted parts
sub part1 {
    my ($workflows, $parts) = @_;
    my $total = 0;

    for my $part (@$parts) {
        if (process_part($workflows, $part)) {
            $total += $part->{x} + $part->{m} + $part->{a} + $part->{s};
        }
    }

    return $total;
}

# Count accepted combinations using range splitting
sub count_accepted {
    my ($workflows, $workflow, $ranges) = @_;

    return 0 if $workflow eq 'R';

    if ($workflow eq 'A') {
        # Count all combinations in current ranges
        my $result = 1;
        for my $attr (qw(x m a s)) {
            my ($lo, $hi) = @{$ranges->{$attr}};
            my $range_size = $hi - $lo + 1;
            $result *= ($range_size > 0 ? $range_size : 0);
        }
        return $result;
    }

    my $total = 0;
    # Make a copy of ranges
    my %ranges_copy = map { $_ => [@{$ranges->{$_}}] } keys %$ranges;

    for my $rule (@{$workflows->{$workflow}}) {
        my ($attr, $op, $value, $destination) = @$rule;

        if (!defined $attr) {
            # Default rule
            $total += count_accepted($workflows, $destination, \%ranges_copy);
        } else {
            my ($lo, $hi) = @{$ranges_copy{$attr}};

            if ($op eq '<') {
                # Split: [lo, value-1] goes to destination, [value, hi] continues
                if ($lo < $value) {
                    # Part that matches the condition
                    my %new_ranges = map { $_ => [@{$ranges_copy{$_}}] } keys %ranges_copy;
                    $new_ranges{$attr} = [$lo, $hi < $value - 1 ? $hi : $value - 1];
                    $total += count_accepted($workflows, $destination, \%new_ranges);
                }
                # Remaining part continues to next rule
                if ($hi >= $value) {
                    $ranges_copy{$attr} = [$lo > $value ? $lo : $value, $hi];
                } else {
                    last;  # No remaining range
                }
            } else {
                # op eq '>'
                # Split: [value+1, hi] goes to destination, [lo, value] continues
                if ($hi > $value) {
                    # Part that matches the condition
                    my %new_ranges = map { $_ => [@{$ranges_copy{$_}}] } keys %ranges_copy;
                    $new_ranges{$attr} = [$lo > $value + 1 ? $lo : $value + 1, $hi];
                    $total += count_accepted($workflows, $destination, \%new_ranges);
                }
                # Remaining part continues to next rule
                if ($lo <= $value) {
                    $ranges_copy{$attr} = [$lo, $hi < $value ? $hi : $value];
                } else {
                    last;  # No remaining range
                }
            }
        }
    }

    return $total;
}

# Part 2: Count all possible accepted combinations
sub part2 {
    my ($workflows) = @_;
    my %initial_ranges = (
        x => [1, 4000],
        m => [1, 4000],
        a => [1, 4000],
        s => [1, 4000],
    );
    return count_accepted($workflows, 'in', \%initial_ranges);
}

# Main
my ($workflows, $parts) = parse_input($input_file);
print "Part 1: ", part1($workflows, $parts), "\n";
print "Part 2: ", part2($workflows), "\n";

File: 2023/day19/ruby/solution.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Day 19: Aplenty - Workflow processing and range analysis

def parse_input(filename)
  text = File.read(filename).strip
  workflow_section, parts_section = text.split("\n\n")

  # Parse workflows
  workflows = {}
  workflow_section.each_line do |line|
    line = line.strip
    name, rules_str = line.split('{')
    rules_str = rules_str.chomp('}')
    rules = rules_str.split(',').map do |rule|
      if rule.include?(':')
        condition, destination = rule.split(':')
        match = condition.match(/([xmas])([<>])(\d+)/)
        attr, op, value = match[1], match[2], match[3].to_i
        { attr: attr, op: op, value: value, destination: destination }
      else
        { attr: nil, op: nil, value: nil, destination: rule }
      end
    end
    workflows[name] = rules
  end

  # Parse parts
  parts = parts_section.each_line.map do |line|
    part = {}
    line.scan(/([xmas])=(\d+)/).each do |attr, value|
      part[attr] = value.to_i
    end
    part
  end

  [workflows, parts]
end

def process_part(workflows, part)
  current = 'in'

  until %w[A R].include?(current)
    workflows[current].each do |rule|
      if rule[:attr].nil?
        current = rule[:destination]
        break
      elsif rule[:op] == '<' && part[rule[:attr]] < rule[:value]
        current = rule[:destination]
        break
      elsif rule[:op] == '>' && part[rule[:attr]] > rule[:value]
        current = rule[:destination]
        break
      end
    end
  end

  current == 'A'
end

def part1(workflows, parts)
  parts.sum do |part|
    if process_part(workflows, part)
      part['x'] + part['m'] + part['a'] + part['s']
    else
      0
    end
  end
end

def count_accepted(workflows, workflow, ranges)
  return 0 if workflow == 'R'

  if workflow == 'A'
    return ranges.values.map { |lo, hi| [0, hi - lo + 1].max }.reduce(1, :*)
  end

  total = 0
  ranges = ranges.dup

  workflows[workflow].each do |rule|
    if rule[:attr].nil?
      total += count_accepted(workflows, rule[:destination], ranges)
    else
      attr = rule[:attr]
      lo, hi = ranges[attr]
      op = rule[:op]
      value = rule[:value]
      destination = rule[:destination]

      if op == '<'
        # Split: [lo, value-1] goes to destination, [value, hi] continues
        if lo < value
          new_ranges = ranges.dup
          new_ranges[attr] = [lo, [hi, value - 1].min]
          total += count_accepted(workflows, destination, new_ranges)
        end
        # Remaining range continues to next rule
        if hi >= value
          ranges[attr] = [[lo, value].max, hi]
        else
          break
        end
      else # op == '>'
        # Split: [value+1, hi] goes to destination, [lo, value] continues
        if hi > value
          new_ranges = ranges.dup
          new_ranges[attr] = [[lo, value + 1].max, hi]
          total += count_accepted(workflows, destination, new_ranges)
        end
        # Remaining range continues to next rule
        if lo <= value
          ranges[attr] = [lo, [hi, value].min]
        else
          break
        end
      end
    end
  end

  total
end

def part2(workflows)
  initial_ranges = {
    'x' => [1, 4000],
    'm' => [1, 4000],
    'a' => [1, 4000],
    's' => [1, 4000]
  }
  count_accepted(workflows, 'in', initial_ranges)
end

def main
  input_file = File.join(__dir__, '..', 'input.txt')
  workflows, parts = parse_input(input_file)
  puts "Part 1: #{part1(workflows, parts)}"
  puts "Part 2: #{part2(workflows)}"
end

main if __FILE__ == $PROGRAM_NAME
